File: packages/pigeon/platform_tests/alternate_language_test_plugin/android/src/test/java/com/example/alternate_language_test_plugin/NullableReturnsTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.alternate_language_test_plugin;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MessageCodec;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

public class NullableReturnsTest {
  @Test
  public void nullArgHostApi() {
    NullableReturns.NullableArgHostApi mockApi = mock(NullableReturns.NullableArgHostApi.class);
    BinaryMessenger binaryMessenger = mock(BinaryMessenger.class);
    NullableReturns.NullableArgHostApi.setUp(binaryMessenger, mockApi);
    ArgumentCaptor<BinaryMessenger.BinaryMessageHandler> handler =
        ArgumentCaptor.forClass(BinaryMessenger.BinaryMessageHandler.class);
    verify(binaryMessenger).setMessageHandler(anyString(), handler.capture());
    MessageCodec<Object> codec = NullableReturns.NullableArgHostApi.getCodec();
    ByteBuffer message =
        codec.encodeMessage(
            new ArrayList<Object>() {
              {
                add(null);
              }
            });
    message.rewind();
    handler
        .getValue()
        .onMessage(
            message,
            (bytes) -> {
              bytes.rewind();
              @SuppressWarnings("unchecked")
              ArrayList<Object> wrapped = (ArrayList<Object>) codec.decodeMessage(bytes);
              assertTrue(wrapped.size() == 1);
            });
  }

  @Test
  public void nullArgFlutterApi() {
    BinaryMessenger binaryMessenger = mock(BinaryMessenger.class);
    doAnswer(
            invocation -> {
              ByteBuffer message = invocation.getArgument(1);
              BinaryMessenger.BinaryReply reply = invocation.getArgument(2);
              message.position(0);
              @SuppressWarnings("unchecked")
              ArrayList<Object> args =
                  (ArrayList<Object>)
                      NullableReturns.NullableArgFlutterApi.getCodec().decodeMessage(message);
              assertNull(args.get(0));
              ByteBuffer replyData =
                  NullableReturns.NullableArgFlutterApi.getCodec()
                      .encodeMessage(
                          new ArrayList<Object>() {
                            {
                              add(args.get(0));
                            }
                          });
              replyData.rewind();
              reply.reply(replyData);
              return null;
            })
        .when(binaryMessenger)
        .send(anyString(), any(), any());
    NullableReturns.NullableArgFlutterApi api =
        new NullableReturns.NullableArgFlutterApi(binaryMessenger);
    boolean[] didCall = {false};
    api.doit(
        null,
        new NullableReturns.NullableResult<Long>() {
          public void success(Long result) {
            didCall[0] = true;
            assertNull(result);
          }

          public void error(Throwable error) {
            assertEquals(error, null);
          }
        });
    assertTrue(didCall[0]);
  }
}

File: packages/pigeon/platform_tests/test_plugin/android/src/test/kotlin/com/example/test_plugin/NullableReturnsTest.kt
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.test_plugin

import io.flutter.plugin.common.BinaryMessenger
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

class NullableReturnsTest {

  @Test
  fun testNullableParameterHost() {
    val binaryMessenger = mockk<BinaryMessenger>(relaxed = true)
    val api = mockk<NullableReturnHostApi>(relaxed = true)

    val output = 1L

    val channelName = "dev.flutter.pigeon.pigeon_integration_tests.NullableReturnHostApi.doit"
    val handlerSlot = slot<BinaryMessenger.BinaryMessageHandler>()

    every { binaryMessenger.setMessageHandler(channelName, capture(handlerSlot)) } returns Unit
    every { api.doit() } returns output

    NullableReturnHostApi.setUp(binaryMessenger, api)

    val codec = PrimitiveHostApi.codec
    val message = codec.encodeMessage(null)
    message?.rewind()
    handlerSlot.captured.onMessage(message) {
      it?.rewind()
      @Suppress("UNCHECKED_CAST") val wrapped = codec.decodeMessage(it) as List<Any>?
      assertNotNull(wrapped)
      wrapped?.let { assertEquals(output, wrapped[0]) }
    }

    verify { binaryMessenger.setMessageHandler(channelName, handlerSlot.captured) }
    verify { api.doit() }
  }

  @Test
  fun testNullableParameterFlutter() {
    val binaryMessenger = mockk<BinaryMessenger>()
    val api = NullableReturnFlutterApi(binaryMessenger)

    val output = 12L

    every { binaryMessenger.send(any(), any(), any()) } answers
        {
          val codec = NullableReturnFlutterApi.codec
          val reply = arg<BinaryMessenger.BinaryReply>(2)
          val replyData = codec.encodeMessage(listOf(output))
          replyData?.position(0)
          reply.reply(replyData)
        }

    var didCall = false
    api.doit {
      didCall = true
      assertEquals(output, it.getOrNull())
    }

    assertTrue(didCall)
  }
}
